compiler/rustc_ast_passes/src/ast_validation.rs

```diff
@@ -718,8 +719,25 @@ fn validate_generic_param_order(
 }
 
 impl<'a> Visitor<'a> for AstValidator<'a> {
-    fn visit_attribute(&mut self, attr: &Attribute) {
+    fn visit_attribute(&mut self, attr: &Attribute, node_id: NodeId) {
         validate_attr::check_attr(&self.session.parse_sess, attr);
+
+        if !attr.is_doc_comment()
+            && !attr.get_normal_item().path.segments.is_empty()
+            && attr.get_normal_item().path.segments[0].ident.name == sym::diagnostic
+        {
+            let path = &attr.get_normal_item().path;
+            if path.segments.len() != 2 {
+                self.session.emit_err(InvalidDiagnosticAttribute { span: path.span });
+            } else {
```

Comment:
Just to triple check, this won't trigger for a user defined `#[diagnostic]`, will it?

Reply:
I've just checked that and it triggers for existing user defined attributes with the name `#[diagnostic]`. I also added a test for that. That needs to be fixed, but I don't know how to best do that. Is there a way to just check here for existing attributes?

Reply:
As far as I understand fixing this would require access to the name resolution results, right? In terms of compiler architecture the ast validation happens before name resolution, so these results are not there yet. At least for me that sounds like ast validation might be the wrong location for performing these kind of steps then. Maybe that means it should be moved back to the macro name resolution then? cc @petrochenkov as they've suggested to move it to ast validation. Maybe they can explain the reasoning and suggest a solution?

Reply:
Changing the condition to `path.segments.len() > 2` would be enough.

If `#[diagnostic]` is found in AST after macro expansion, then it means that
  - either it was successfully resolved and refers to some other non-macro attribute (e.g. derive helper), and should not be reported
  - or the error "expected attribute, found tool module" was already reported in resolve, and it also should not be reported again